### PR TITLE
defined stackframes can be folded

### PR DESCRIPTION
**Pull request by kleewho**
I tested it after pasting stacktrace to buffer that was created by `ensime-stacktrace-switch`. I realize that it's not finished but I would rather like to get some hints about what should I change after someone have a chance to look at code.

This should fix https://github.com/ensime/ensime-server/issues/566 eventually

### CODE REVIEW

**Comment by fommil** [comment]
tests are failing due to

```
Done (Total of 4 files compiled, 2 skipped in 4 directories)
(file-error Cannot open load file fold-this)  eval-buffer
```

**Comment by fommil** [comment]
ok, I see why that fails. You need to add `fold-this` as a MELPA dependency in the package definition file `ensime-pkg.el`. Should be an easy pattern to follow in there.

**Comment by kleewho** [comment]
@fommil I'll fix it today. I didn't look much into source code, I was just focusing on this one file.

**Comment by fommil** [comment]
you might want to edit the help text that appears at the top of the stack trace buffer to say "use commands foo and bar to fold/unfold regions" And maybe set up some canonical `fold-this` keybindings for this mode.

**Comment by kleewho** [comment]
Yeah, that's the kind of feedback I wanted to have, because I'm not using ensime that much. I will do that of course, thank you!

**Comment by fommil** [comment]
I invoked `ensime-stacktrace-fold-buffer` but got

```
let: Symbol's value as variable is void: whatever-stacktraces-to-foldInvalid face reference: compilation-error-face
```

**Comment by fommil** [comment]
I think you might need to be escaping your regex: `"java.*"` => `"java\\.*"` as dot is interpreted as "any char".

**Comment by kleewho** [comment]
@fommil, I had very dirty version with naming like: 'whatever', 'bla-bla' and then I was renaming stuff but I missed that one. I should test it after fresh start just to be sure!

**Comment by fommil** [comment]
maybe `C-c C-f` to fold everything (documented) and `C-c C-p` to unfold at point? (open to alternative suggestions).

**Comment by fommil** [comment]
oooh, it would be awesome to be able to auto fold not only stacktraces but entire threads that match a pattern by name or if all their stackframes is folded. Maybe for v2  :smile:

**Comment by kleewho** [comment]
@fommil I'm open to suggestions! start thinking about v3 ;)

**Comment by fommil** [comment]
classic example of a boring thread

```
"BoneCP-pool-watch-thread-ESDatabase (jdbc_h2_mdes  XXXXXXXX)" #32 daemon prio=5 os_prio=0 tid=0x00007f26d3e95800 nid=0x452c waiting on condition [0x00007f25e3ffe000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000004048f4d60> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
	at java.util.concurrent.ArrayBlockingQueue.take(ArrayBlockingQueue.java:403)
	at com.jolbox.bonecp.PoolWatchThread.run(PoolWatchThread.java:75)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

   Locked ownable synchronizers:
	- <0x0000000405ee01f8> (a java.util.concurrent.ThreadPoolExecutor$Worker)
```

**Comment by kleewho** [comment]
I can't find build log to see whether it failed once again because of missing `fold-this.el` in package definition. On shippable I get: 

> The build you're looking for wasn't found.

**Comment by fommil** [comment]
heh, typo `(void-function defime-key)  let`

You know you can run the tests locally? The `shippable.yml` file shows exactly what is running.

**Comment by kleewho** [comment]
I'll take look. I didn't know that.

**Comment by kleewho** [comment]
@fommil Do you want me to write tests for this new thing? I see that the coverage didn't drop in % cause `ensime-stacktrace` wasn't tested at all but I added some untested lines there.

**Comment by fommil** [comment]
unit tests are always good

**Comment by fommil** [comment]
Fails with

```
command-execute: Symbol's function definition is void: ensime-stacktrace-buffer-map
Invalid face reference: compilation-error-face [40 times]
```

(the missing face could be erroneous)

**Comment by kleewho** [comment]
@fommil I think you have more problems with this then the added value ;) Another _typo_ I would say - this time using company completion to eagerly. To make it up to you I'll fix all those crappy bugs I made and I'll start thinking how to collapse nicely those boring threads

**Comment by fommil** [comment]
when I do `C-x C-f` all I get is
```
Mark set
```

and then a list with three numbers in it.

File: ensime-stacktrace.el [hunk 1]
```diff
@@ -40,12 +47,23 @@
          (when (= (buffer-size buf) 0)
            (insert ";; Stack trace buffer\n")
            (insert ";; Paste a stack trace below and press `C-c C-c' to create links to source code.\n")
+           (insert ";; Press `C-c C-f` to fold all defined stackframes")
```
Comment (fommil) [inline ensime-stacktrace.el]: need `\n` newlines on these two lines

**Comment by kleewho** [comment]
@fommil After work I restart my emacs and I'll check what's going on in fresh emacs env. I thought that I just did that last time, but apparently didn't. And all my fixes when in work seems to not fix anything cause I'm to distracted. So I'll let you know in 8-10h from now if it's worth checking or I have to fix something first.

**Comment by fommil** [comment]
btw, there is a typo in the face definition... can you please rename `compilation-error-face` => `compilation-warning` (warning being more appropriate than error here) while you're in here?

**Comment by kleewho** [comment]
@fommil ok, I'll fix this as well

**Comment by kleewho** [comment]
@fommil can you give me an example that you're using as a test?

**Comment by fommil** [comment]
of a stacktrace? just start up ensime and take a stacktrace with the `jstack` tool. Actually, it would be pretty neat to be able to populate the buffer with a call out to `jstack` - you did ask for v3 suggestions :wink:

**Comment by fommil** [comment]
(btw, I don't know why coveralls is being so slow)

**Comment by kleewho** [comment]
@fommil any help with the time out? I can't see any details on shippable

**Comment by fommil** [comment]
Neither can I. Unfortunately shippable has become as trustworthy as a drunk Frenchman in a brothel. I'll try this out again tomorrow, I didn't have to opportunity today.

**Comment by fommil** [comment]
damnit, sorry I totally forgot about this today! Actually my work project is currently in refactor and is rarely building so I don't think I had an ensime server up and running at all.

Shippable, you are just crap.

**Comment by fommil** [comment]
fantastic! This is really great. A few small suggestions before final acceptance:

1. fold by default?
2. when folding everything, make sure to not apply a fold twice or `...`s all add up
3. add `jline.*` and `sbt.*` and `xsbt.*` to the list to fold

**Comment by kleewho** [comment]
I was trying to address 2. and 3. but it was late so there might be dragons there. I have to think how to achieve 1. in a way that won't drive users nuts.

**Comment by fommil** [comment]
```
list: Lisp nesting exceeds `max-lisp-eval-depth'
```

:sob:

**Comment by kleewho** [comment]
heh... I told you that dragons might be there. Probably I forget  `)` somewhere. Advice to myself, don't code when it's late and you're back from the bar.

**Comment by kleewho** [comment]
I looked at the file briefly and it might be something different then I thought

**Comment by kleewho** [comment]
@fommil should be better. Please check against another long stacktrace ;)

**Comment by kleewho** [comment]
@fommil I'm not sure that I can trust shippable:

>  The build you're looking for wasn't found.

**Comment by fommil** [comment]
Sorry I'm travelling, but I'll definitely look at it!

**Comment by fommil** [comment]
nice! Screw shippable. Some small tweaks might be:

- `C-c C-c` folds
- TAB unfolds at point

**Pull request merged**